### PR TITLE
Feature: add custom id claim

### PR DESCRIPTION
**Pull request by gigovich**
## Describe your changes
This feature allows using the custom claim in the JWT token as a user ID.

## Issue ticket number and link
https://github.com/netbirdio/netbird/issues/507

### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [x] Is a feature enhancement
- [x] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: management/server/jwtclaims/extractor.go [hunk 1]
```diff
@@ -14,51 +15,81 @@ const (
...
+	return ce
+}
+
+// FromToken extracts claims from the token (after auth)
+func (c *ClaimsExtractor) FromToken(token *jwt.Token) AuthorizationClaims {
 	claims := token.Claims.(jwt.MapClaims)
 	jwtClaims := AuthorizationClaims{}
-	jwtClaims.UserId = claims[UserIDClaim].(string)
-	accountIdClaim, ok := claims[authAudience+AccountIDSuffix]
+	jwtClaims.UserId = claims[c.userIDClaim].(string)
```
Comment (mlsmaycon) [inline management/server/jwtclaims/extractor.go]: What is the behavior when a claim doesn't exist in a JWT?

File: management/cmd/management.go [hunk 2]
```diff
@@ -178,8 +179,13 @@ var (
 				tlsEnabled = true
 			}
 
-			httpAPIHandler, err := httpapi.APIHandler(accountManager, config.HttpConfig.AuthIssuer,
-				config.HttpConfig.AuthAudience, config.HttpConfig.AuthKeysLocation, appMetrics)
+			httpApiAuthCfg := httpapi.AuthCfg{
+				Issuer:       config.HttpConfig.AuthIssuer,
+				Audience:     config.HttpConfig.AuthAudience,
+				UserIDClaim:  config.HttpConfig.AuthUserIDClaim,
```
Comment (mlsmaycon) [inline management/cmd/management.go:185]: This value will be empty on upgrade as it didn't exist before. We can add a small check and set it to "sub"

File: infrastructure_files/setup.env.example [hunk 3]
```diff
@@ -7,6 +7,8 @@ NETBIRD_AUTH_OIDC_CONFIGURATION_ENDPOINT=""
 NETBIRD_AUTH_AUDIENCE=""
 # e.g. netbird-client
 NETBIRD_AUTH_CLIENT_ID=""
+# if you want to use a custom claim for the user ID instead of 'sub', set it here
+# NETBIRD_AUTH_USER_ID_CLAIM=""
```
Comment (mlsmaycon) [inline infrastructure_files/setup.env.example:11]: We could set the value to sub already here or on the base.setup.env

File: management/server/grpcserver.go [hunk 4]
```diff
@@ -200,7 +213,7 @@ func (s *GRPCServer) registerPeer(peerKey wgtypes.Key, req *proto.LoginRequest)
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "invalid jwt token, err: %v", err)
 		}
-		claims := jwtclaims.ExtractClaimsWithToken(token, s.config.HttpConfig.AuthAudience)
+		claims := s.jwtClaimsExtractor.FromToken(token) // jwtclaims.ExtractClaimsWithToken(token, s.config.HttpConfig.AuthAudience)
```
Comment (mlsmaycon) [inline management/server/grpcserver.go]: comment leftover

File: management/server/http/api/types.gen.go [hunk 5]
```diff
@@ -574,6 +574,9 @@ type User struct {
 
 	// Status User's status
 	Status UserStatus `json:"status"`
+
+	// IsCurrent is true if authenticated user is the same as this user
+	IsCurrent bool `json:"is_current"`
```
Comment (mlsmaycon) [inline management/server/http/api/types.gen.go]: We are missing the specification update

File: management/server/jwtclaims/extractor.go (hunk 1 above)
Comment (gigovich) [reply management/server/jwtclaims/extractor.go]: I added checks in review commit, to fix places where we created account and didn't check is userID empty or not.

File: management/cmd/management.go (hunk 2 above)
Comment (gigovich) [reply management/cmd/management.go:185]: This property initialized in the Claims Extractor, and if UserIDClaim is empty, we use `sub`.
https://github.com/gigovich/netbird/blob/feat-add-custom-id-claim/management/server/jwtclaims/extractor.go#L62-L62

File: infrastructure_files/setup.env.example (hunk 3 above)
Comment (gigovich) [reply infrastructure_files/setup.env.example:11]: As I mentioned, it set in the code if it is empty.

File: management/server/http/api/types.gen.go (hunk 5 above)
Comment (gigovich) [reply management/server/http/api/types.gen.go]: Fixed

File: management/server/grpcserver.go (hunk 4 above)
Comment (gigovich) [reply management/server/grpcserver.go]: Fixed

**Review by mlsmaycon — APPROVED** [review]

**Pull request merged**